authority/provisioner/sign_ssh_options.go

```diff
@@ -223,15 +221,49 @@ func (m *sshCertificateValidityModifier) Modify(cert *ssh.Certificate) error {
 		cert.ValidBefore = uint64(t.Add(d).Unix())
 	}
 
-	diff := time.Duration(cert.ValidBefore-cert.ValidAfter) * time.Second
-	switch {
-	case diff < min:
-		return errors.Errorf("ssh certificate duration cannot be lower than %s", min)
-	case diff > max:
-		return errors.Errorf("ssh certificate duration cannot be greater than %s", max)
-	default:
+	return nil
+}
+
+type sshProvCredValidityModifier struct {
+	*Claimer
+	// Remaining duration on the provisioning credential.
+	// E.g. x5c provisioners use a certificate as a provisioning credential.
+	// That certificate should not be able to provision new certificates with
+	// a duration longer than the remaining duration on the provisioning
+	// certificate.
+	rem time.Duration
+}
+
+// Modify has the following procedure:
+//  1) Use the existing ValidBefore & ValidAfter values if they are set; else
+//  2.a) Use the default duration to set the validity bounds and
+//  2.b) If 'rem' is non-zero and less than the default duration then
+//       reduce the duration to 'rem'.
+//
+// Step 2) only applies if ValidBefore is not set in the request.
+func (m *sshProvCredValidityModifier) Modify(cert *ssh.Certificate) error {
+	defMod := &sshValidityModifier{m.Claimer}
+	if err := defMod.Modify(cert); err != nil {
+		return err
+	}
+	if m.rem == 0 {
 		return nil
 	}
+
+	if cert.ValidBefore < cert.ValidAfter {
+		return errors.New("ssh certificate validBefore cannot be before validAfter")
+	}
+	dur, err := time.ParseDuration(fmt.Sprintf("%ds", cert.ValidBefore-cert.ValidAfter))
+	if err != nil {
+		return errors.Wrapf(err, "error converting ssh certificate duration; "+
+			"vaf: %d, vbf: %d, d: %d", cert.ValidAfter, cert.ValidBefore, cert.ValidBefore-cert.ValidAfter)
+	}
+	if m.rem < dur {
+		t := time.Unix(int64(cert.ValidAfter), 0)
+		cert.ValidBefore = uint64(t.Add(m.rem).Unix())
+	}
```

Comment:
I think we can do like in the `profileProvCredDuration` where you apply the `profileDefaultDuration` of the lower value. But later you need to check that the duration is less than m.rem too in case that was the user who set it.